core/src/identity.rs

```diff
@@ -0,0 +1,195 @@
+
+pub mod ed25519;
+#[cfg(not(any(target_os = "emscripten", target_os = "unknown")))]
+pub mod rsa;
+#[cfg(feature = "secp256k1")]
+pub mod secp256k1;
+
+pub mod error;
+
+use self::error::*;
+// use log::debug;
+use crate::{PeerId, keys_proto};
+
+/// Identity keypair of the local node.
+///
+/// # Example: Generating RSA keys with OpenSSL
+///
+/// ```text
+/// openssl genrsa -out private.pem 2048
+/// openssl pkcs8 -in private.pem -inform PEM -topk8 -out private.pk8 -outform DER -nocrypt
+/// rm private.pem      # optional
+/// ```
+///
+/// Loading the keys:
+///
+/// ```ignore
+/// let keypair = Keypair::rsa_from_pkcs8(include_bytes!("private.pk8"));
+/// ```
+///
+#[derive(Clone)]
+pub enum Keypair {
+    Ed25519(ed25519::Keypair),
+    #[cfg(not(any(target_os = "emscripten", target_os = "unknown")))]
+    Rsa(rsa::Keypair),
+    #[cfg(feature = "secp256k1")]
+    Secp256k1(secp256k1::Keypair)
+}
+
+impl Keypair {
+    pub fn new_ed25519() -> Keypair {
+        Keypair::Ed25519(ed25519::Keypair::generate())
+    }
+
+    #[cfg(feature = "secp256k1")]
+    pub fn new_secp256k1() -> Keypair {
+        Keypair::Secp256k1(secp256k1::Keypair::generate())
+    }
+
+    #[cfg(not(any(target_os = "emscripten", target_os = "unknown")))]
+    pub fn rsa_from_pkcs8(pkcs8_der: &[u8]) -> Result<Keypair, DecodingError> {
+        rsa::Keypair::from_pkcs8(pkcs8_der).map(Keypair::Rsa)
+    }
+
+    /// Decode a keypair from a Secp256k1 secret key in an ECPrivateKey
+    /// structure as defined in [RFC5915] in DER encoding.
+    ///
+    /// [RFC5915]: https://tools.ietf.org/html/rfc5915
+    #[cfg(feature = "secp256k1")]
+    pub fn secp256k1_from_der(sk_der: &[u8]) -> Result<Keypair, DecodingError> {
+        secp256k1::Keypair::from_secret_der(sk_der).map(Keypair::Secp256k1)
+    }
+
+    pub fn sign(&self, msg: &[u8]) -> Result<Vec<u8>, SigningError> {
+        use Keypair::*;
+        match self {
+            Ed25519(ref pair) => Ok(pair.sign(msg)),
+            #[cfg(not(any(target_os = "emscripten", target_os = "unknown")))]
+            Rsa(ref pair) => pair.sign(msg),
+            #[cfg(feature = "secp256k1")]
+            Secp256k1(ref pair) => Ok(pair.sign(msg)),
+        }
+    }
+
+    pub fn public(&self) -> PublicKey {
+        use Keypair::*;
+        match self {
+            Ed25519(pair) => PublicKey::Ed25519(pair.public()),
+            #[cfg(not(any(target_os = "emscripten", target_os = "unknown")))]
+            Rsa(pair) => PublicKey::Rsa(pair.public()),
+            #[cfg(feature = "secp256k1")]
+            Secp256k1(pair) => PublicKey::Secp256k1(pair.public()),
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum PublicKey {
+    Ed25519(ed25519::PublicKey),
+    #[cfg(not(any(target_os = "emscripten", target_os = "unknown")))]
+    Rsa(rsa::PublicKey),
+    #[cfg(feature = "secp256k1")]
+    Secp256k1(secp256k1::PublicKey)
+}
+
+impl PublicKey {
+    pub fn verify(&self, msg: &[u8], sig: &[u8]) -> bool {
```

Comment:
Maybe `Result<(), ()>` is more appropriate?

Reply:
In what sense more appropriate?

Reply:
Do you mean the implied semantics of `Ok` and `Err` are superior to `true` and `false`? If you think that is preferable, I don't mind.

Reply:
> Do you mean the implied semantics of `Ok` and `Err` are superior to `true` and `false`?

That's what I mean, yes. I don't have a strong opinion though.